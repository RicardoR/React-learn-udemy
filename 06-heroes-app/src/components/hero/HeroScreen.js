import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroById } from '../../helpers/getHeroById';

export const HeroScreen = () => {
  const { heroId } = useParams();
  const navigate = useNavigate();
  const hero = getHeroById(heroId);
  const handleReturn = () => {
    navigate(-1);
  };

  if (!hero) {
    return <Navigate to="/" />;
  }

  const { id, superhero, alter_ego, publisher, first_appearance, characters } =
    hero;

  const imgPath = `/assets/${id}.jpg`;

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={imgPath}
          alt={superhero}
          className="img-fluid img-thumbnail"
        />
      </div>
      <div className="col-8">
        <h3>{superhero}</h3>
        <ul className="list-group">
          <li className="list-group-item">
            <strong>Alter ego:</strong> {alter_ego}
          </li>
          <li className="list-group-item">
            <strong>Publisher:</strong> {publisher}
          </li>
          <li className="list-group-item">
            <strong>First appearance:</strong> {first_appearance}
          </li>
        </ul>
        <h4 className="mt-3">Characters</h4>
        <p>{characters}</p>
        <button className="btn btn-outline-info" onClick={handleReturn}>
          Regresar
        </button>
      </div>
    </div>
  );
};
