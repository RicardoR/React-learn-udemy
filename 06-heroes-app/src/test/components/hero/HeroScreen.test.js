import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { HeroScreen } from '../../../components/hero/HeroScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('HeroScreen Ut', () => {
  test('should not render an hero screen without an hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <Routes>
          <Route path="/hero" element={<HeroScreen />} />
          <Route path="/" element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(HeroScreen)).toHaveLength(0);
    expect(wrapper.find('h1').text().trim()).toBe('No hero page');
  });

  test('should render the hero screen when param exists and hero has found', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-thor']}>
        <Routes>
          <Route path="/hero/:heroId" element={<HeroScreen />} />
          <Route path="/" element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(HeroScreen)).toHaveLength(1);
    expect(wrapper.find('h3').text().trim()).toBe('Thor');
  });

  test('should return to back screen', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-thor']}>
        <Routes>
          <Route path="/hero/:heroId" element={<HeroScreen />} />
        </Routes>
      </MemoryRouter>
    );
    wrapper.find('button').prop('onClick')();
    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });

  test('should display no hero page when hero does not exists', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/no-existe']}>
        <Routes>
          <Route path="/hero/:heroId" element={<HeroScreen />} />
          <Route path="/" element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );
    expect(wrapper.find(HeroScreen)).toHaveLength(0);
    expect(wrapper.find('h1').text().trim()).toBe('No hero page');
  });
});
