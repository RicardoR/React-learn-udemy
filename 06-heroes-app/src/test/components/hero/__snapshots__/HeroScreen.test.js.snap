// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroScreen Ut should not render an hero screen without an hero 1`] = `ReactWrapper {}`;

exports[`HeroScreen Ut should render the hero screen when param exists and hero has found 1`] = `ReactWrapper {}`;
